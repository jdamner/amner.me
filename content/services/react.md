---
title: Javascript & React
---

I've worked with [React](https://reactjs.org) and [React Native](https://reactnative.dev) for a number of years and find the stateful nature of the React framework to really work well in my head. I love the component based approach and the ability to create reusable components that can be used in multiple places.

This site is build using React and the [NextJS](https://www.nextjs.org/) framework. I parse [markdown](https://daringfireball.net/projects/markdown/) files for the content. This site is totally open source too, so if you're curious about how it's built, you can check out the [source code](https://www.github.com/jdamner/amner.me).

Also, as a statically generated site, it's super fast and easy to deploy.
