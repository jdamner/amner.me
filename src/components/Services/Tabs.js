import React from "react";
import Seperator from "../Global/Seperator";
import TabButton from "./TabButton";
import TabContent from "./TabContent";
import { StaticQuery, graphql } from "gatsby";
import { GatsbyImage, getImage } from "gatsby-plugin-image";

export default function Tabs() {

    const [activeTab, setActiveTab] = React.useState(0);
    const [tabOpen, setTabOpen] = React.useState(false);

    const handleTabChange = (index) => {
        setTabOpen(index === activeTab ? !tabOpen : true);
        setActiveTab(index);
    }

    return (
        <StaticQuery
            query={graphql`
            query MyQuery {
                allFile(filter: {sourceInstanceName: {eq: "services"}}) {
                  edges {
                    node {
                      id
                      childrenMarkdownRemark {
                        frontmatter {
                          title
                          thumbnail {
                            childImageSharp {
                              gatsbyImageData(
                                width: 200
                                placeholder: BLURRED
                                formats: [AUTO, WEBP, AVIF]
                              )
                            }
                          }
                        }
                        html
                      }
                    }
                  }
                }
              }`}
            render={data => {
                const { allFile } = data;
                const TabButtons = allFile.edges.map(edge => {

                    const { frontmatter } = edge.node.childrenMarkdownRemark[0];
                    if (!frontmatter) {
                        return null;
                    }

                    const file = frontmatter.thumbnail ? getImage(frontmatter.thumbnail) : null;
                    console.log(file, frontmatter);
                    const image = file ? <GatsbyImage image={file} alt={frontmatter.title} objectFit={'contain'} /> : null;
                    

                    return (
                        <TabButton
                            key={edge.node.id}
                            name={frontmatter.title}
                            image={image}
                            active={tabOpen && activeTab === edge.node.id}
                            index={edge.node.id}
                            onClick={() => handleTabChange(edge.node.id)}
                        />
                    );
                })

                const active = allFile.edges.find(edge => edge.node.id === activeTab)
                const tab = active ? active.node.childrenMarkdownRemark[0] : null;


                return (
                    <div className="tab">
                        <div className='blog-wrap-title'>
                            <Seperator />
                            <h2>Special Skills</h2>
                            <Seperator modifier='rotate' />
                        </div>
                        <div className="tab-buttons-wrapper">
                            {TabButtons}
                        </div>
                        <div className="tab-content-wrapper">
                            {tabOpen ? <TabContent tab={tab} /> : null}
                        </div>
                    </div>
                )
            }
            }
        />
    )
}