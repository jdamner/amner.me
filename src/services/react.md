---
title: React
thumbnail: ../assets/react-logo.png
---

Working with SQL in raw queries, and using frameworks such as Eloquent and WP_Query.
Coaching and supporting a team of developers grow their skills and meet their objectives
Supporting Project Managers and Business Development teams for technical guidance on quoting and supporting project lifecycles.
Debugging and maintaining PHP projects built in many frameworks, including bespoke PHP frameworks developed in house.
Developing systems that interact with complex data, such as data provided via third-party tools and services that run on device and output in raw formats.