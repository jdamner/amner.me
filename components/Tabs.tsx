"use client";

import React from "react";
/* API */
import { useState } from "react";

/* Components */
import TabButton from "./Atoms/Links/TabButton";

import { useAllServices } from "../utils";

export default function Tabs({
  tabs,
  defaultContent,
  title,
}: {
  tabs: ReturnType<typeof useAllServices>;
  defaultContent: React.ReactNode;
  title: React.ReactNode;
}) {
  const [activeTab, setActiveTab] = useState(0);
  const [tabOpen, setTabOpen] = useState(false);

  const handleTabChange = (index: number) => {
    setTabOpen(index === activeTab ? !tabOpen : true);
    setActiveTab(index);
  };

  const TabButtons = tabs.map((tab, index) => {
    const active = tabOpen && activeTab === index;
    const ariaCurrent = activeTab === index && tabOpen ? "page" : undefined;
    return (
      <TabButton
        key={index}
        delay={index * 0.25}
        role="tab"
        id={`tab-${index}`}
        aria-current={ariaCurrent}
        aria-controls={`tab-${index}`}
        active={active}
        onClick={() => handleTabChange(index)}
      >
        {tab.data.title}
      </TabButton>
    );
  });

  const contentClassName =
    "prose prose-slate dark:prose-invert mr-auto text-slate-900 dark:text-slate-200";

  function ActiveTab() {
    const tab = tabs[activeTab];
    return (
      <div
        className={contentClassName}
        id={`tab-${activeTab}`}
        role="tabpanel"
        aria-labelledby={`tab-${activeTab}`}
      >
        {tab.content}
      </div>
    );
  }

  function NonActiveTab() {
    return <div className={contentClassName}>{defaultContent}</div>;
  }

  return (
    <>
      {title}
      <div className="md:grid md:grid-cols-3" role="tablist">
        <div className="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
          {TabButtons}
        </div>
        <div className="md:col-span-2">
          {tabOpen ? <ActiveTab /> : <NonActiveTab />}
        </div>
      </div>
    </>
  );
}
