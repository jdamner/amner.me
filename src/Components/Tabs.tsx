'use client'
import React from "react";
/* API */
import { useState } from "react";

/* Components */
import TabButton from "../Atoms/Links/TabButton";
import ReactMarkdown from 'react-markdown';

export default function Tabs({ tabs, defaultContent, title }:
  {
    tabs: {
      title: string,
      content: string
    }[],
    defaultContent: string,
    title: React.ReactNode
  }
) {

  const [activeTab, setActiveTab] = useState(0);
  const [tabOpen, setTabOpen] = useState(false);

  const handleTabChange = (index: number) => {
    setTabOpen(index === activeTab ? !tabOpen : true);
    setActiveTab(index);
  }

  const TabButtons = tabs.map((tab, index) => {
    const active = tabOpen && activeTab === index || !tabOpen;
    const ariaCurrent = activeTab === index && tabOpen ? 'page' : undefined;
    return (
      <TabButton
        key={index}
        delay={index * 0.25}
        role="tab"
        id={`tab-${index}`}
        aria-current={ariaCurrent}
        aria-controls={`tab-${index}`}
        active={active}
        onClick={() => handleTabChange(index)}
      >{tab.title}</TabButton>
    );
  });

  const contentClassName = 'prose prose-slate dark:prose-invert mr-auto text-slate-900 dark:text-slate-200'

  const ActiveTab = () => {
    const tab = tabs[activeTab];
    return (
      <div
        className={contentClassName}
        id={`tab-${activeTab}`}
        role="tabpanel"
        aria-labelledby={`tab-${activeTab}`}
      >
        <ReactMarkdown
          components={
            { a: (props) => <a {...props} target="_blank" rel="noopener noreferrer" /> }
          }>{tab.content}</ReactMarkdown>
      </div>
    )
  }


  const NonActiveTab = () => {
    return (
      <ReactMarkdown className={contentClassName}>{defaultContent}</ReactMarkdown>
    )
  }

  return (
    <>
      {title}
      <div className="md:grid md:grid-cols-3" role="tablist">
        <div className="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
          {TabButtons}
        </div>
        <div className="md:col-span-2">
          {tabOpen ? <ActiveTab /> : <NonActiveTab />}
        </div>
      </div>
    </>
  )
}
