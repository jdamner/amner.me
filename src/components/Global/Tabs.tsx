import { useState } from "react";
import { event } from "../../api/Insights";

import type { post } from "../../types/post.type";

import TabButton from "../Links/TabButton";
import Title from "./Title";
import ReactMarkdown from 'react-markdown';


export default function Tabs({ tabs, content }: { tabs: post[], content: string }): JSX.Element {

  const [activeTab, setActiveTab] = useState(0);
  const [tabOpen, setTabOpen] = useState(false);

  const handleTabChange = (index) => {
    event('tab-change', { index, tabOpen });
    setTabOpen(index === activeTab ? !tabOpen : true);
    setActiveTab(index);
  }

  const TabButtons = tabs.map((tab, index) => {
    const active = tabOpen && activeTab === index || !tabOpen;
    return (
      <TabButton
        key={index}
        index={index + 1}
        name={tab.title}
        active={active}
        onClick={() => handleTabChange(index)}
      />
    );
  });

  const contentClassName = 'prose prose-slate dark:prose-invert mr-auto px-3 md:px-0 text-slate-900 dark:text-slate-200'

  const ActiveTab = () => {
    const tab = tabs[activeTab];
    return (
      <ReactMarkdown className={contentClassName}>{tab.content}</ReactMarkdown>
    )
  }


  const NonActiveTab = () => {
    return (
      <ReactMarkdown className={contentClassName}>{ content }</ReactMarkdown>
    )
  }

  return (
    <section className="container mx-auto my-5 py-5 px-3 md:px-0" id="services-tabs">
      <Title title={'Techincal Skills'}>Web Development</Title>
      <div className="md:grid md:grid-cols-3">
        <div className="flex flex-row md:flex-col flex-wrap md:flex-nowrap gap-3 md:gap-0 items-start md:items-end pr-5 mb-3">
          {TabButtons}
        </div>
        <div className="md:col-span-2">
          {tabOpen ? <ActiveTab /> : <NonActiveTab />}
        </div>
      </div>
    </section>
  )
}
