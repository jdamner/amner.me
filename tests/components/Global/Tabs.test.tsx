import { render } from "@testing-library/react"
import Tabs from "../../../src/components/Global/Tabs"
import { getTestPosts } from "../../helper"

describe("Tabs", () => {
  it("renders correctly", () => {
    const { container } = render(<Tabs tabs={ getTestPosts() } content='<p>Test Content</p>' />)
    expect(container).toMatchSnapshot()
  })
})
