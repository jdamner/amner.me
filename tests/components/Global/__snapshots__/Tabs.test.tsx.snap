// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders correctly 1`] = `
<div>
  <div
    class="container mx-auto my-5 py-5 px-3 md:px-0"
  >
    <div
      class="flex flex-col md:flex-row md:items-end mb-5"
    >
      <h2
        class="text-lg font-sans font-semibold tracking-tight text-left font-black uppercase text-xl md:text-6xl break-words basis-1/3"
      >
        Techincal Skills
      </h2>
      <div
        class="w-full text-left uppercase text-2xl font-black"
      >
        <div
          class="mb-2"
        >
          Web Development
        </div>
        <div
          class="bg-black h-2 w-full"
          style="max-width: 0px;"
        />
      </div>
    </div>
    <div
      class="md:grid md:grid-cols-3"
    >
      <div
        class="flex flex-row md:flex-col items-start md:items-end pr-4 mr-3 mb-3 md:mb-0"
      >
        <button
          aria-current="true"
          class="text-xl md:text-3xl font-bold text-gray-900"
          tabindex="0"
        >
          Test Post
        </button>
        <button
          aria-current="true"
          class="text-xl md:text-3xl font-bold text-gray-900"
          tabindex="1"
        >
          Test Post
        </button>
        <button
          aria-current="true"
          class="text-xl md:text-3xl font-bold text-gray-900"
          tabindex="2"
        >
          Test Post
        </button>
      </div>
      <div
        class="col-span-2"
      >
        <div
          class="mr-auto prose"
          content="<p>Test Content</p>"
        >
          <p>
            Test Content
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
